src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -0,0 +1,293 @@
+    {
+        this.client = client;
+        this.pubSubName = pubSubName;
+        this.topicName = topicName;
+        this.options = options;
+        this.messageHandler = handler;
+    }
+
+    /// <summary>
+    /// Dynamically subscribes to messages on a PubSub topic provided by the Dapr sidecar.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing messages provided by the sidecar.</returns>
+    public async Task SubscribeAsync(CancellationToken cancellationToken = default)
+    {
+        //Prevents the receiver from performing the subscribe operation more than once (as the multiple initialization messages would cancel the stream).
+        if (hasInitialized)
+            return;
+        hasInitialized = true;
+
+        var stream = await GetStreamAsync(cancellationToken);
+
+        //Retrieve the messages from the sidecar and write to the messages channel
+        var fetchMessagesTask = FetchDataFromSidecarAsync(stream, topicMessagesChannel.Writer, cancellationToken);
+
+        //Process the messages as they're written to either channel
+        var acknowledgementProcessorTask = ProcessAcknowledgementChannelMessagesAsync(cancellationToken);
+        var topicMessageProcessorTask = ProcessTopicChannelMessagesAsync(cancellationToken);
+
+        try
+        {
+            await Task.WhenAll(fetchMessagesTask, acknowledgementProcessorTask, topicMessageProcessorTask);
+        }
+        catch (OperationCanceledException)
+        {
+            await DisposeAsync();
+        }
+    }
+
+    /// <summary>
+    /// Retrieves or creates the bidirectional stream to the DaprClient for transacting pub/sub subscriptions.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>The stream connection.</returns>
+    private async Task<AsyncDuplexStreamingCall<P.SubscribeTopicEventsRequestAlpha1, P.SubscribeTopicEventsResponseAlpha1>> GetStreamAsync(CancellationToken cancellationToken)
+    {
+        await semaphore.WaitAsync(cancellationToken);
+
+        try
+        {
+            return clientStream ??= client.SubscribeTopicEventsAlpha1(cancellationToken: cancellationToken);
+        }
+        finally
+        {
+            semaphore.Release();
+        }
+    }
+
+    /// <summary>
+    /// Acknowledges the indicated message back to the Dapr sidecar with an indicated behavior to take on the message.
+    /// </summary>
+    /// <param name="messageId">The identifier of the message the behavior is in reference to.</param>
+    /// <param name="behavior">The behavior to take on the message as indicated by either the message handler or timeout message handling configuration.</param>
+    /// <returns></returns>
+    private async Task AcknowledgeMessageAsync(string messageId, TopicResponseAction behavior)
```

Comment:
Nit: this should also take a `CancellationToken` argument?  (But probably the original one passed to `ProcessTopicChannelMessagesAsync()`, not the linked one as we shouldn't consider time writing the acknowledgement as part of the time given to the user's handler.

Reply:
Might as well - added